#2715 add child NOTES.txt rendering

jgleonard opened the pull request:
This PR should address the rendering of NOTES.txt from subcharts.

https://github.com/kubernetes/helm/issues/2751

rajatjindal committed:
fix lint warning

Signed-off-by: jgleonard <jgleonard@gmail.com>

Commit:
replace with a link to the latest releases page

This removes a step required every time we release Helm, making it
simpler to cut a new release.

Signed-off-by: jgleonard <jgleonard@gmail.com>

AdamDang committed:
Update capabilities.go

Signed-off-by: jgleonard <jgleonard@gmail.com>

diclophis commented:
This would be super helpful

bacongobbler commented:
Is there a way we can gate this behind a feature flag? Many charts probably don't want their respective subcharts to emit their notes as they're being used to supplement the parent chart and the user doesn't interact with it (e.g. wordpress depending on mariadb).

Thanks!

bacongobbler commented:
passed to the helm command itself, like `helm install --render-subchart-notes` or something

vainikkaj commented:
Great to see this moving forward.
I'll be happy to help testing this.

jgleonard commented:
Thanks @vainikkaj !

I had it working locally and then of course needed to merge since it was an out of date branch.  Now I've got conflicts and the build broke.  Working through it.

jgleonard committed:
add child NOTES.txt rendering

Signed-off-by: jgleonard <jgleonard@gmail.com>

jgleonard committed:
fix(helm): add --render-subchart-notes flag to 'helm install' and 'helm upgrade'

When 'helm <install|upgrade> --render-subchart-notes ...' is run, this will include
the notes from the subchart when rendered via Tiller.

Closes #2751

Signed-off-by: jgleonard <jgleonard@gmail.com>

jgleonard committed:
Merge branch 'master' into 2751

jgleonard committed:
make docs

Signed-off-by: jgleonard <jgleonard@gmail.com>

jgleonard committed:
gofmt -s

Signed-off-by: jgleonard <jgleonard@gmail.com>

jgleonard commented:
@vainikkaj and @bacongobbler Should be all clear now

jgleonard commented:
`✔-1 ~/go/src/k8s.io/helm [2751|✔]
15:18 $ bin/helm install -n notestest ~/testdata/testing
NAME:   notestest
LAST DEPLOYED: Thu Nov  1 15:18:24 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/Service
NAME                TYPE       CLUSTER-IP     EXTERNAL-IP  PORT(S)  AGE
notestest-subchart  ClusterIP  10.99.245.31   <none>       80/TCP   0s
notestest-testing   ClusterIP  10.96.120.147  <none>       80/TCP   0s

==> v1beta2/Deployment

NAME                DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
notestest-subchart  1        0        0           0          0s
notestest-testing   1        1        1           0          0s

==> v1/Pod(related)

NAME                                 AGE
notestest-subchart-767466565c-4dxkm  0s
notestest-testing-dc79d59c-kml9p     0s


NOTES:
PARENT

✔ ~/go/src/k8s.io/helm [2751|✔]
15:18 $ bin/helm upgrade notestest ~/testdata/testing --render-subchart-notes
Release "notestest" has been upgraded. Happy Helming!
LAST DEPLOYED: Thu Nov  1 15:18:34 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/Service
NAME                TYPE       CLUSTER-IP     EXTERNAL-IP  PORT(S)  AGE
notestest-subchart  ClusterIP  10.99.245.31   <none>       80/TCP   10s
notestest-testing   ClusterIP  10.96.120.147  <none>       80/TCP   10s

==> v1beta2/Deployment

NAME                DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
notestest-subchart  1        1        1           1          10s
notestest-testing   1        1        1           1          10s

==> v1/Pod(related)

NAME                                 AGE
notestest-subchart-767466565c-4dxkm  10s
notestest-testing-dc79d59c-kml9p     10s


NOTES:
PARENT

CHILD`

vainikkaj commented:
Cool, I'll give it a spin during the weekend and post my findings here.

bacongobbler commented on pkg/helm/option.go:
```diff
@@ -339,6 +341,20 @@ func InstallReuseName(reuse bool) InstallOption {
+// InstallSubNotes will (if true) instruct Tiller to render SubChart Notes
+func InstallSubNotes(enable bool) InstallOption {
+	return func(opts *options) {
+		opts.subNotes = enable
```
by making a call to `InstallSubNotes`, it is also instructing tiller to render subcharts' notes during a `helm upgrade`. It'd be better to move the `subNotes` variable into the installReq and upgradeReq structs so they can be isolated similar to `upgradeDescription`.

bacongobbler replied:
From Helm's perspective, behaviorally speaking nothing will change. From the SDK's point of view, it blocks this  from happening:

```
res, err := client.InstallReleaseFromChart(
    ...,
    helm.InstallSubNotes(true)
)
...
res, err := client.UpgradeReleaseFromChart(
    ...
)
```

which will cause the subcharts to render when `client.UpgradeReleaseFromChart` is called.

Does that make sense?

jgleonard replied:
This should be resolve with latest push.

jgleonard committed:
formatting

Signed-off-by: jgleonard <jgleonard@gmail.com>

jgleonard committed:
fold subNotes into updateReq and instReq

Signed-off-by: jgleonard <jgleonard@gmail.com>

bacongobbler reviewed (approved):
LGTM! This has been marked as a `size/L` though so it'll need a sign-off from another maintainer before it can be merged.

vainikkaj commented:
Sorry for being late to the party.
I've tried this against different stable charts in Minikube and it works as expected.

One thing that I observed is that when chart contains multiple subcharts then NOTES.txt gets IMO rather cluttered.
Let's use stable/prometheus-operator as an example:

```
NOTES:
1. Get your 'admin' user password by running:

   kubectl get secret --namespace monitoring prometheus-operator-grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo

2. The Grafana server can be accessed via port 80 on the following DNS name from within your cluster:

   prometheus-operator-grafana.monitoring.svc.cluster.local

   From outside the cluster, the server URL(s) are:
     http://grafana.minikube


3. Login with the password from step 1 and the username: admin
#################################################################################
######   WARNING: Persistence is disabled!!! You will lose your data when   #####
######            the Grafana pod is terminated.                            #####
#################################################################################

The Prometheus Operator has been installed. Check its status by running:
  kubectl --namespace monitoring get pods -l "release=prometheus-operator"

Visit https://github.com/coreos/prometheus-operator for instructions on how
to create & configure Alertmanager and Prometheus instances using the Operator.

1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus-node-exporter,release=prometheus-operator" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:80

kube-state-metrics is a simple service that listens to the Kubernetes API server and generates metrics about the state of the objects.
The exposed metrics can be found here:
https://github.com/kubernetes/kube-state-metrics/tree/master/Documentation#documentation.

The metrics are exported on the HTTP endpoint /metrics on the listening port.
In your case, prometheus-operator-kube-state-metrics.monitoring.svc.cluster.local:8080/metrics

They are served either as plaintext or protobuf depending on the Accept header.
They are designed to be consumed either by Prometheus itself or by a scraper that is compatible with scraping a Prometheus client endpoint.
```

And an alternative format where chart names are included in NOTES:

```
NOTES:
prometheus-operator/charts/kube-state-metrics

kube-state-metrics is a simple service that listens to the Kubernetes API server and generates metrics about the state of the objects.
The exposed metrics can be found here:
https://github.com/kubernetes/kube-state-metrics/tree/master/Documentation#documentation.

The metrics are exported on the HTTP endpoint /metrics on the listening port.
In your case, prometheus-operator-kube-state-metrics.monitoring.svc.cluster.local:8080/metrics

They are served either as plaintext or protobuf depending on the Accept header.
They are designed to be consumed either by Prometheus itself or by a scraper that is compatible with scraping a Prometheus client endpoint.


prometheus-operator/charts/grafana

1. Get your 'admin' user password by running:

   kubectl get secret --namespace monitoring prometheus-operator-grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo

2. The Grafana server can be accessed via port 80 on the following DNS name from within your cluster:

   prometheus-operator-grafana.monitoring.svc.cluster.local

   From outside the cluster, the server URL(s) are:
     http://grafana.minikube


3. Login with the password from step 1 and the username: admin
#################################################################################
######   WARNING: Persistence is disabled!!! You will lose your data when   #####
######            the Grafana pod is terminated.                            #####
#################################################################################

prometheus-operator/charts/prometheus-node-exporter

1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace monitoring -l "app=prometheus-node-exporter,release=prometheus-operator" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:80

prometheus-operator

The Prometheus Operator has been installed. Check its status by running:
  kubectl --namespace monitoring get pods -l "release=prometheus-operator"

Visit https://github.com/coreos/prometheus-operator for instructions on how
to create & configure Alertmanager and Prometheus instances using the Operator.
```
What do you guys think? For me this is an UX improvement and can definitely be done in different PR later.

vainikkaj commented:
@bacongobbler Any chance of getting this into 2.12?

jgleonard commented:
@vainikkaj Thanks for checking it out and good catch on the formatting, might be reasonable to open that up as a new issue with "good first issue" tag.

Clecompt commented:
Will this be in 2.13?

michelleN commented on pkg/helm/option.go:
```diff
@@ -53,7 +53,7 @@ type options struct {
 	// if set, skip CRD hook only
 	disableCRDHook bool
-	// name of release
+	// if set, render SubChart Notes
```
did you mean to make this comment change?

jgleonard replied:
Oops!  No, there had been a var there before that was removed.  I'll revert that line.

michelleN reviewed (commented):
I tried testing with the `stable/wordpress`chart which has a child mariadb chart with notes but using the flag `--render-subchart-notes` did not render the notes for mariadb. Am I doing something wrong?

UPDATE: This was something on my end. I just didn't have the right setup! I'm re-testing now!

michelleN reviewed (commented):
This looks great @jgleonard. I just have that one question about the comment and then I'd be happy to LGTM. Thank you.

jgleonard committed:
Restore comment text

Signed-off-by: Jon Leonard <jon.leonard@sanofi.com>

jgleonard commented:
Looks like this one needs porting to v3, not sure how to add a label...

hickeyma commented:
@jgleonard Thanks for raising this port.

I have ported it in PR #6512

dudicoco commented:
In which Helm version is this feature included?

hickeyma commented:
@dudicoco it is in both Helm 3.0+ and 2.13.0+

dudicoco commented:
@hickeyma i'm using 2.13.1 and the child chart's NOTES.txt is not being rendered.
Any ideas?

jgleonard commented:
@dudioco are you using the command line option ?

On Wed, Dec 11, 2019 at 11:33 AM dudicoco <notifications@github.com> wrote:

> @hickeyma <https://github.com/hickeyma> i'm using 2.13.1 and the child
> chart's NOTES.txt is not being rendered.
> Any ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/helm/helm/pull/4088?email_source=notifications&email_token=AGLFEDPXELSMD7CSDRWJBITQYEI5PA5CNFSM4FAUL3X2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGTYIEY#issuecomment-564626451>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLFEDITAJ4LOXN4MJ2UB6LQYEI5PANCNFSM4FAUL3XQ>
> .
>

dudicoco commented:
@jgleonard I have now tried the --render-subchart-notes flag and I get an unknown flag error. It also does not appear under helm -h.

hickeyma commented:
@dudicoco Is the flag available in `helm install` or `helm upgrade`?

dudicoco commented:
@hickeyma it's working now with the flag, I accidentally ran it using helm template.
Thanks!
